karapace/schema_reader.py

```diff
@@ -254,105 +266,122 @@ def handle_messages(self):
                 if self.ready and add_offsets:
                     self.offset_watcher.offset_seen(self.offset)
 
-    def handle_msg(self, key: dict, value: dict):
-        if key["keytype"] == "CONFIG":
-            if "subject" in key and key["subject"] is not None:
-                if not value:
-                    self.log.info("Deleting compatibility config completely for subject: %r", key["subject"])
-                    self.subjects[key["subject"]].pop("compatibility", None)
-                    return
-                self.log.info(
-                    "Setting subject: %r config to: %r, value: %r", key["subject"], value["compatibilityLevel"], value
-                )
-                if not key["subject"] in self.subjects:
-                    self.log.info("Adding first version of subject: %r with no schemas", key["subject"])
-                    self.subjects[key["subject"]] = {"schemas": {}}
-                subject_data = self.subjects.get(key["subject"])
-                subject_data["compatibility"] = value["compatibilityLevel"]
-            else:
-                self.log.info("Setting global config to: %r, value: %r", value["compatibilityLevel"], value)
-                self.config["compatibility"] = value["compatibilityLevel"]
-        elif key["keytype"] == "SCHEMA":
+    def _handle_msg_config(self, key: dict, value: Optional[dict]) -> None:
+        subject = key.get("subject")
+        if subject is not None:
+            if subject not in self.subjects:
+                self.log.info("Adding first version of subject: %r with no schemas", subject)
+                self.subjects[subject] = {"schemas": {}}
+
             if not value:
-                subject, version = key["subject"], key["version"]
-                self.log.info("Deleting subject: %r version: %r completely", subject, version)
-                if subject not in self.subjects:
-                    self.log.error("Subject %s did not exist, should have", subject)
-                elif version not in self.subjects[subject]["schemas"]:
-                    self.log.error("Version %d for subject %s did not exist, should have", version, subject)
-                else:
-                    self.subjects[subject]["schemas"].pop(version, None)
-                return
```

Comment:
Here we have a memory leak. This could be the last use of the schema and `self.schemas` should be cleared too. This PR is _not_ fixing this, that was done on https://github.com/aiven/karapace/pull/284 and had to be reverted https://github.com/aiven/karapace/pull/295

Reply:
I believe there is another bug here, but I'm not sure how to fix this.

- If the very last produced message is for a hard delete of a schema
- The topic is compacted
- The server is restarted

Then the information of the global ID is lost, because the `key` contains only the `version` of the schema in the subject, but not the `id` that is global to all subjects. That means the `id` will be reused.